dtox.sh

```diff
@@ -4,41 +4,69 @@ set -euo pipefail
 
 ROOT="$(git rev-parse --show-toplevel)"
 
+BASE_MODE="${BASE_MODE:-build}"
+CACHE_MODE="${CACHE_MODE:-}"
+CACHE_TAG="${CACHE_TAG:-latest}"
+
 BASE_INPUT=(
   "${ROOT}/docker/base/Dockerfile"
+  "${ROOT}/docker/base/install_pythons.sh"
 )
 base_hash=$(cat "${BASE_INPUT[@]}" | git hash-object -t blob --stdin)
 
-function base_id() {
-  docker images -q -f label=base_hash="${base_hash}" pantsbuild/pex:base
+function base_image_id() {
+  docker image ls -q "ghcr.io/pantsbuild/pex/base:${base_hash}"
 }
 
-if [[ -z "$(base_id)" ]]; then
+if [[ "${BASE_MODE}" == "build" && -z "$(base_image_id)" ]]; then
   docker build \
-    --tag pantsbuild/pex:base \
-    --label base_hash="${base_hash}" \
+    --tag ghcr.io/pantsbuild/pex/base:latest \
+    --tag "ghcr.io/pantsbuild/pex/base:${base_hash}" \
     "${ROOT}/docker/base"
+elif [[ "${BASE_MODE}" == "pull" ]]; then
+  docker pull "ghcr.io/pantsbuild/pex/base:${base_hash}"
 fi
 
 USER_INPUT=(
   "${ROOT}/docker/user/Dockerfile"
   "${ROOT}/docker/user/create_docker_image_user.sh"
 )
 user_hash=$(cat "${USER_INPUT[@]}" | git hash-object -t blob --stdin)
-if [[ -z "$(docker images -q -f label=user_hash="${user_hash}" pantsbuild/pex:user)" ]]; then
+
+function user_image_id() {
+  docker image ls -q "pantsbuild/pex/user:${user_hash}"
+}
+
+if [[ -z "$(user_image_id)" ]]; then
   docker build \
-    --build-arg BASE_ID="$(base_id)" \
+    --build-arg BASE_IMAGE_TAG="${base_hash}" \
     --build-arg USER="$(id -un)" \
     --build-arg UID="$(id -u)" \
     --build-arg GROUP="$(id -gn)" \
     --build-arg GID="$(id -g)" \
-    --tag pantsbuild/pex:user \
-    --label user_hash="${user_hash}" \
+    --tag pantsbuild/pex/user:latest \
+    --tag "pantsbuild/pex/user:${user_hash}" \
     "${ROOT}/docker/user"
 fi
 
+if [[ "${CACHE_MODE}" == "pull" ]]; then
```

Comment:
This is actually the only bit that has to do with the new cache data-only image, the other changes in this script support interop between local build and seeding with `*_MODE=pull` as well as test fixes (altering a few volume mount paths).

Reply:
And this is a pretty particular trick I had to fight a bit for. I'll add a comment about the necessity of an empty volume + a run to cause the local volume to be filled from the data image.